### PR TITLE
ARROW-17517: [C++] Test engine API in public API test

### PR DESCRIPTION
**Pull request by pitrou**
Also some assorted header inclusion cleanups.

Followup to PR #13965.

### CODE REVIEW

**Comment by pitrou** [comment]
This is a crude way to solve the issue and some more elaborate solution could probably be pushed. @westonpace

File: cpp/src/arrow/engine/api.h [hunk 1]
```diff
@@ -19,6 +19,4 @@
 
 #pragma once
 
-#include "arrow/engine/substrait/extension_set.h"
 #include "arrow/engine/substrait/extension_types.h"
-#include "arrow/engine/substrait/serde.h"
```
Comment (westonpace) [inline cpp/src/arrow/engine/api.h:24]: Why remove `serde.h`?  Is it transitively pulling in something internal or was this change a preference thing?  This is probably the most significant public header so I'll need to figure out how we can get it in this list.

Comment (pitrou) [reply cpp/src/arrow/engine/api.h:24]: It is transitively pulling some internals. As I said the solution here deserved to be improved :-)

Comment (westonpace) [reply cpp/src/arrow/engine/api.h:24]: Ok, I'll make an attempt to clean up exec & engine soon but we should keep going with this PR as-is.  Thanks for noticing this.

**Comment by pitrou** [comment]
@westonpace @aucahuasi Some engine maintainer should probably take this task up.

File: cpp/src/arrow/engine/api.h (hunk 1 above)
Comment (westonpace) [reply cpp/src/arrow/engine/api.h:24]: https://github.com/apache/arrow/pull/14131

**Comment by pitrou** [comment]
@westonpace I updated this to just contain the test additions and a few cleanups.

**Review by westonpace — APPROVED** [review]
Thanks for the further cleanup.

**Comment by pitrou** [comment]
The LLVM deprecation errors are unrelated.

**Pull request merged**